### PR TITLE
fix(xss): fix 4 context classification edge cases (#7086)

### PR DESCRIPTION
**Pull request by Gengyscan**
## Fix: 4 XSS Context Analyzer Edge Cases

Fixes #7086

### Changes

Based on the XSS context analyzer from #7076, this PR fixes all 4 misclassification edge cases:

#### 1. javascript: URIs in href/action/formaction
Attribute values starting with javascript: are now classified as ContextScript instead of ContextAttribute.

#### 2. JSON script blocks
<script type="application/json"> and similar non-executable types are no longer flagged as ContextScript. Added isExecutableScriptType() helper that defers the inScript decision until after reading the 	ype attribute.

#### 3. Case-insensitive marker detection  
The fast-path exit (strings.Contains) now uses strings.ToLower for both body and marker, preventing missed reflections when case differs.

#### 4. srcdoc attribute context
srcdoc attributes are classified as ContextHTMLText since they contain nested HTML that gets parsed by the browser.

### Testing
- 4 new test cases in context_test.go, one per edge case
- All existing tests preserved

### Files
- pkg/fuzz/analyzers/xss/context.go — Core fixes in DetectReflections tokenizer
- pkg/fuzz/analyzers/xss/context_test.go — 4 new tests
- pkg/fuzz/analyzers/xss/analyzer.go — Analyzer integration (from #7076)
- pkg/fuzz/analyzers/xss/types.go — Type definitions (from #7076)
- pkg/fuzz/analyzers/analyzers.go — Analyzer registry (from #7076)
- pkg/protocols/http/*.go — HTTP protocol integration (from #7076)

## Summary by CodeRabbit

* **New Features**
  * Added an XSS context analyzer ("xss_context") that detects reflected markers across HTML (tags, attributes, scripts, comments) and replays context‑appropriate payloads to confirm issues.

* **Improvements**
  * Analyzer now records response body, headers, and status for richer results.
  * Improved thread‑safety for random utilities used during concurrent fuzzing.
  * Ensured analyzer parameters are initialized to avoid nil‑map issues.
  * Enhanced request/fuzz integration and payload handling.

* **Tests**
  * Added extensive tests and benchmarks for XSS detection and context handling.

### CODE REVIEW

File: pkg/fuzz/analyzers/xss/context.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+						}
+
+						if isEventHandler(attrName) {
+							// Event handler attributes contain JavaScript
+							ctx = ContextScript
+						}
+
+						if strings.HasPrefix(strings.ToLower(strings.TrimSpace(attrVal)), "javascript:") {
+							ctx = ContextScript
+						}
```
[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/context.go]: Good catch! Addressed in commit 2ae3b64. Added `isScriptURLAttribute()` gating with a whitelist of navigation-target URL attributes (href, action, formaction, src, xlink:href, data, poster, codebase). Non-executable sinks like `data-*` and `aria-*` are no longer promoted to `ContextScript`.

File: pkg/fuzz/analyzers/xss/context.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+	if normalized == "" {
+		return true
+	}
+
+	switch normalized {
+	case "text/javascript", "text/ecmascript", "application/javascript", "application/ecmascript", "module":
+		return true
+	default:
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/context.go:228]: Addressed in commit 2ae3b64. The function now:
1. Extracts the MIME essence by stripping parameters after `;` (so `text/javascript; charset=utf-8` correctly matches)
2. Returns `false` for invalid essences (e.g., `;charset=utf-8` where stripping leaves an empty string)
3. Includes all 16 JavaScript MIME type essences per the WHATWG MIME Sniffing spec + `module`

Thanks for the thorough spec research!

**Comment by Gengyscan** [comment]
Pushed commit 2ae3b64 addressing all CodeRabbit findings plus two additional issues we discovered:

**CodeRabbit findings addressed:**
1. **`detectAttrQuoting` word boundary** (Minor) — ✅ Already fixed in d8171af
2. **`javascript:` URI gating** (Major) — ✅ Added `isScriptURLAttribute()` whitelist
3. **`isExecutableScriptType` MIME essence** (Major) — ✅ Full 16-essence WHATWG whitelist + proper essence extraction

**Additional fixes:**
4. **`detectCharacterSurvival` false negatives** (Critical) — Characters were concatenated (`canary<>"'/`) so only `<` was ever independently detectable. Now each character is interleaved with its own canary prefix.
5. **Template literal payload** (High) — `selectPayloads` for `ContextScriptString` now includes `${alert(1)}` for backtick template literal injection.

All fixes include corresponding test cases.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+// detectCharacterSurvival checks which XSS-critical characters survived server-side encoding
+func detectCharacterSurvival(body string, canary string) CharacterSet {
+	return CharacterSet{
+		LessThan:     strings.Contains(body, canary+"<"),
+		GreaterThan:  strings.Contains(body, canary+">"),
+		DoubleQuote:  strings.Contains(body, canary+`"`),
+		SingleQuote:  strings.Contains(body, canary+`'`),
+		ForwardSlash: strings.Contains(body, canary+"/"),
+	}
+}
```
[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/analyzer.go:231]: Respectfully, this is a false positive. The interleaved canary pattern **does** work correctly.

The generated string is: `canary<canary>canary"canary'canary/`

For example with canary=`ABC`: `ABC<ABC>ABC"ABC'ABC/`

`detectCharacterSurvival` checks:
- `strings.Contains(body, canary+"<")` — matches at position 0 (`ABC<`) ✅
- `strings.Contains(body, canary+">")` — matches at position 4 (`ABC>`) ✅
- `strings.Contains(body, canary+'"')` — matches at position 8 (`ABC"`) ✅
- `strings.Contains(body, canary+"'")` — matches at position 12 (`ABC'`) ✅
- `strings.Contains(body, canary+"/")` — matches at position 16 (`ABC/`) ✅

Each character is **preceded** by its own canary in the interleaved string, so `canary + char` is independently findable for all 5 characters. This was the whole point of the interleaving fix — the old code concatenated all chars after a single canary (`canary<>"'/`), making only `<` detectable.

File: pkg/fuzz/analyzers/xss/types.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+var scriptURLAttributes = map[string]struct{}{
+	"href":       {},
+	"action":     {},
+	"formaction": {},
+	"src":        {},
+	"xlink:href": {},
+	"data":       {},
+	"poster":     {},
+	"codebase":   {},
+}
```
Comment (Gengyscan) [inline pkg/fuzz/analyzers/xss/types.go]: Good catch! Removed `poster` from `scriptURLAttributes` in commit ad292c5. You're right that modern browsers treat `poster` as an image resource URL and don't execute `javascript:` URIs there — including it caused false positives.

The remaining attributes (`href`, `action`, `formaction`, `src`, `xlink:href`, `data`, `codebase`) are all genuine navigation/execution targets where `javascript:` URIs can fire.

File: pkg/fuzz/analyzers/xss/analyzer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/analyzer.go:231]: Respectfully, this is a false positive. The interleaved canary pattern **does** work correctly.

The generated string is: `canary<canary>canary"canary'canary/`

For example with canary=`ABC`: `ABC<ABC>ABC"ABC'ABC/`

`detectCharacterSurvival` checks:
- `strings.Contains(body, canary+"<")` — matches at position 0 (`ABC<`) ✅
- `strings.Contains(body, canary+">")` — matches at position 4 (`ABC>`) ✅
- `strings.Contains(body, canary+'"')` — matches at position 8 (`ABC"`) ✅
- `strings.Contains(body, canary+"'")` — matches at position 12 (`ABC'`) ✅
- `strings.Contains(body, canary+"/")` — matches at position 16 (`ABC/`) ✅

Each character is **preceded** by its own canary in the interleaved string, so `canary + char` is independently findable for all 5 characters. This was the whole point of the interleaving fix — the old code concatenated all chars after a single canary (`canary<>"'/`), making only `<` detectable.

File: pkg/fuzz/analyzers/xss/types.go (hunk 4 above)
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/types.go]: Good catch! Removed `poster` from `scriptURLAttributes` in commit ad292c5. You're right that modern browsers treat `poster` as an image resource URL and don't execute `javascript:` URIs there — including it caused false positives.

The remaining attributes (`href`, `action`, `formaction`, `src`, `xlink:href`, `data`, `codebase`) are all genuine navigation/execution targets where `javascript:` URIs can fire.

Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/types.go]: Good follow-up! You're right — `codebase` is an obsolete base-URL resolver, not a navigation target. Removed it in commit f9e577c.

The final `scriptURLAttributes` map is now:
- `href` — `<a>`, `<area>`, `<base>`
- `action` / `formaction` — `<form>` and form override attributes
- `src` — `<iframe>`, `<embed>` (keeping this broad since `<iframe src=javascript:>` is a genuine vector)
- `xlink:href` — SVG navigation
- `data` — `<object>` (borderline but historically exploitable, keeping for scanner coverage)

Thanks for the thorough analysis with the MDN and Firefox bug references!

File: pkg/fuzz/analyzers/xss/context.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+						if unquoted {
+							ctx = ContextAttributeUnquoted
+						}
+
+						if isEventHandler(attrName) {
+							// Event handler attributes contain JavaScript
+							ctx = ContextScript
+						}
+
+						if isScriptURLAttribute(attrName) && strings.HasPrefix(strings.ToLower(strings.TrimSpace(attrVal)), "javascript:") {
```
[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit 5a0fac6. The URI scheme detection now:

1. Normalizes attribute values by stripping internal ASCII whitespace (tabs, newlines, CR, form feed) before prefix checking
2. Detects `data:` URIs with executable MIME types (`text/html`, `image/svg+xml`) in script-URL attributes and classifies them as `ContextScript`
3. Non-executable data: URIs like `data:image/png` are correctly left as `ContextAttribute`

Tests added for both positive and negative cases.

[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit 5a0fac6. Added `normalizeURIScheme()` helper that strips internal control whitespace (`\t`, `\n`, `\r`, `\f`) before prefix matching. This matches browser behavior where URI scheme parsing ignores these characters.

Tests added for both tab and newline obfuscated `javascript:` variants.

[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit 5a0fac6. The scheme check now includes `vbscript:` alongside `javascript:`. While IE is deprecated, enterprise environments with IE compatibility modes still exist, so a security scanner should detect this vector.

Test added for `vbscript:` in href attribute.

**Comment by Gengyscan** [comment]
Pushed commit 5a0fac6 addressing all 3 Neo security review findings:

**1. `data:` URI XSS bypass** (Critical) — Now detects `data:text/html` and `data:image/svg+xml` URIs in navigation-target attributes as `ContextScript`.

**2. Tab/newline injection bypass** (Major) — Added `normalizeURIScheme()` that strips internal ASCII whitespace (`\t`, `\n`, `\r`, `\f`) before scheme prefix matching, matching browser URL parsing behavior.

**3. `vbscript:` scheme** (Minor) — Added detection for legacy IE `vbscript:` URIs.

All fixes include tests (6 new test cases: 3 positive, 1 negative for data:, 2 whitespace variants).

File: pkg/fuzz/analyzers/xss/context.go [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+							ctx = ContextAttributeUnquoted
+						}
+
+						if isEventHandler(attrName) {
+							// Event handler attributes contain JavaScript
+							ctx = ContextScript
+						}
+
+						if isScriptURLAttribute(attrName) {
+							normalizedVal := normalizeURIScheme(attrVal)
```
[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/context.go:87]: Addressed in commit 5136b78. The `normalizeURIScheme()` function now applies `url.PathUnescape()` to the prefix before the colon, decoding percent-encoded schemes like `%6A%61%76%61%73%63%72%69%70%74:` into `javascript:` before the HasPrefix check. This handles both percent-encoded `javascript:` and `data:` URIs.

Only the scheme prefix (before `:`) is decoded to avoid altering the URI payload.

File: pkg/fuzz/analyzers/xss/context.go [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+							if strings.HasPrefix(normalizedVal, "javascript:") || strings.HasPrefix(normalizedVal, "vbscript:") {
+								ctx = ContextScript
+							} else if strings.HasPrefix(normalizedVal, "data:") {
+								// data:text/html and data:image/svg+xml can execute scripts
+								if strings.Contains(normalizedVal, "text/html") || strings.Contains(normalizedVal, "image/svg+xml") {
```
[the remark this answers is not in the document]
Comment (Gengyscan) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit 5136b78. Added `application/xhtml+xml` to the executable MIME type checks for `data:` URIs. The detection now covers `text/html`, `image/svg+xml`, and `application/xhtml+xml` — all MIME types that browsers execute JavaScript in when loaded via iframe/object/embed elements.

**Comment by Gengyscan** [comment]
@coderabbitai resume

**Comment by Gengyscan** [comment]
All 4 findings from the latest CodeRabbit review (commit d826c01) have been addressed:

**analyzer.go** — commit 66cf8d95832e
1. **Early exit case-insensitive**: `strings.ToLower` applied to both `body` and `canary` before the `strings.Contains` check.
2. **`detectCharacterSurvival` case-insensitive**: introduced `bodyL`/`canaryL` lowercase vars; all five character checks (`<`, `>`, `"`, `'`, `/`) now use them.

**context.go** — commit c554dbceb0cd
3. **`data:` URI MIME parsing**: media-type is extracted before the first `,` or `;`, preventing false positives when the payload data portion contains a MIME string like `text/html`.
4. **`detectAttrQuoting` whitespace**: the search loop now skips optional whitespace between the attribute name and `=`, so `href = "val"` is handled correctly.

> **Minor (CSP)**: Properly threading replay response headers through `replayAndVerify` requires a broader refactor; tracked as a follow-up to keep this PR focused.

**Comment by Gengyscan** [comment]
Addressed all remaining CodeRabbit findings in commit `abf46b5e`:

**Already fixed** (from previous commits, CR comments referenced stale code):
- `isExecutableScriptType` — full MIME whitelist + param stripping via `;` split ✓
- `normalizeURIScheme` — whitespace re-stripped after percent-decode ✓
- Case-insensitive canary — `strings.ToLower` on both early exit and `detectCharacterSurvival` ✓
- `data:` URI — media-type parsed before `,`/`;` only ✓
- `detectAttrQuoting` — word boundary check + whitespace tolerance around `=` ✓

**Fixed in this commit:**
- **`analyzer.go:143`** — moved `defer` restore immediately after `SetValue`, before `Rebuild()`. A `Rebuild()` failure no longer leaks the mutated component state.
- **`context.go:25`** — removed dead `inRCDATA` variable (written but never read) and unused `rcdataElements` map from `types.go`.

All tests pass (`go test ./pkg/fuzz/... -timeout 120s`).

**Comment by Mzack9999** [comment]
Closing this PR as the edge cases reported in Issue #7086 targeted the XSS context analyzer from PR #7076, which was not merged. The XSS context analyzer has been implemented via PR #7164 which correctly handles all reported edge cases by design. This PR also includes extensive reformatting of core HTTP protocol files. Thank you for your contribution.
